node/subsystem-util/src/nesting_sender.rs

```diff
@@ -0,0 +1,148 @@
+//! provide a sender and dispatch received messages to the root module - it is completely
+//! irrelevant how complex that module is, it might consist of child modules also having the need
+//! to spawn and receive messages, which in turn do the same, still the root logic stays unchanged.
+//! Everything is isolated to the level where it belongs, while we still keep a single task scope
+//! in all non blocking/not CPU intensive parts, which allows us to share data via references for
+//! example.
+//!
+//! Because the wrapping is optional and transparent to the lower modules, each module can also be
+//! used at the top directly without any wrapping, e.g. for standalone use or for testing purposes.
+//!
+//! In the interest of time I refrain from providing a usage example here for now, but would instead
+//! like to point you to the dispute-distribution subsystem which makes use of this architecture.
+//!
+//! Nothing is ever for free of course: Each level adds an indirect function call to message
+//! sending. which should be cheap enough for most applications, but something to keep in mind. In
+//! particular we avoided the use of of async traits, which would have required memory allocations
+//! on each send. Also cloning of `NestingSender` is more expensive than cloning a plain
+//! mpsc::Sender, the overhead should be negligible though.
+//!
+//! Further limitations: Because everything is routed to the same channel, it is not possible with
+//! this approach to put back pressure on only a single source (as all are the same). If a module
+//! has a task that requires this, it indeed has to spawn a long running task which can do the
+//! back-pressure on that message source or we make it its own subsystem. This is just one of the
+//! situations that justifies the complexity of asynchronism.
+
+use std::{convert::identity, sync::Arc};
+
+use futures::{channel::mpsc, SinkExt};
+
+/// A message sender that supports nested messages.
+///
+/// This sender wraps an `mpsc::Sender` and a conversion function for converting given messages of
+/// type `M1` to the message type actually supported by the mpsc (`M`).
+pub struct NestingSender<M, M1> {
+	sender: mpsc::Sender<M>,
+	conversion: Arc<dyn Fn(M1) -> M + 'static + Send + Sync>,
+}
+
+impl<M> NestingSender<M, M>
+where
+	M: 'static,
+{
+	/// Create a new "root" sender.
+	///
+	/// This is a sender that directly passes messages to the mpsc.
+	pub fn new_root(root: mpsc::Sender<M>) -> Self {
+		Self { sender: root, conversion: Arc::new(identity) }
+	}
+}
+
+impl<M, M1> NestingSender<M, M1>
+where
+	M: 'static,
+	M1: 'static,
+{
+	/// Create a new `NestingSender` which wraps a given "parent" sender.
+	///
+	/// Doing the necessary conversion from the child message type `M1` to the parent message type
+	/// `M2.
+	/// M1 -> M2 -> M
+	/// Inputs:
+	///		F(M2) -> M (via parent)
+	///		F(M1) -> M2 (via child_conversion)
+	/// Result: F(M1) -> M
```

Comment:
The formatting here is a bit messed up. I tried to fix it.

```suggestion
	/// This takes care of the necessary conversion from the child message type `M1` to the parent
	/// message type `M2`.
	///
	/// ```text
	/// M1 -> M2 -> M
	/// Inputs:
	///		F(M2) -> M (via parent)
	///		F(M1) -> M2 (via child_conversion)
	/// Result: F(M1) -> M
	/// ```
```